src/Microsoft.ServiceHub.Framework/Container/GlobalBrokeredServiceContainer.cs

```diff
@@ -269,6 +269,23 @@ public IServiceBroker GetSecureServiceBroker(ServiceActivationOptions options)
 	/// </devremarks>
 	public ValueTask<TraceSource?> GetTraceSourceForBrokeredServiceAsync(IServiceBroker serviceBroker, ServiceMoniker serviceMoniker, ServiceActivationOptions options, bool clientRole, CancellationToken cancellationToken) => this.GetTraceSourceForConnectionAsync(serviceBroker, serviceMoniker, options, clientRole, cancellationToken);
 
+	/// <summary>
+	/// Applies typical transformations on a descriptor fro brokered service clients and services.
+	/// </summary>
+	/// <param name="descriptor">The stock descriptor used for this service.</param>
+	/// <param name="serviceBroker">A service broker that may be used to acquire other, related services as necessary to mutate the <paramref name="descriptor"/>.</param>
+	/// <param name="serviceActivationOptions">The activation options for the service.</param>
+	/// <param name="clientRole">A value indicating whether the <paramref name="descriptor"/> is about to be used to activate a client proxy or client connection; use <see langword="false" /> when activating the service itself.</param>
+	/// <param name="cancellationToken">A cancellation token.</param>
+	/// <returns>The modified descriptor.</returns>
+	internal async ValueTask<ServiceRpcDescriptor> ApplyDescriptorSettingsAsync(ServiceRpcDescriptor descriptor, IServiceBroker serviceBroker, ServiceActivationOptions serviceActivationOptions, bool clientRole, CancellationToken cancellationToken)
+	{
+		TraceSource? traceSource = await this.GetTraceSourceForConnectionAsync(serviceBroker, descriptor.Moniker, serviceActivationOptions, clientRole, cancellationToken).ConfigureAwait(false);
+		return descriptor
+			.WithJoinableTaskFactory(this.joinableTaskFactory)
```

Comment:
Would it be a very micro-optimization to have a single method with both args to avoid cloning twice?

Reply:
Ya. And our established pattern doesn't allow for it. If C# records had been available when we first created these descriptors I would probably have used them since cloning and the `with` keyword is built-in. 
If this becomes a perf problem, we can define our own `With` method that takes multiple arguments. That's a complex thing to get right, but [I've done it before](https://devblogs.microsoft.com/premier-developer/immutableobjectgraph-has-some-big-updates-now-and-more-coming-soon/), so we can do it here if needed.